File: src/core/system/library/schema.rb

require 'enso'

module Schema
  
  def self.object_key(obj)
    obj[obj.schema_class.key.name]
  end
  
  #run DFS on obj's parent hierarchy and return first non-nil result
  #similar to find() from Ruby's Enumerable API
  def self.lookup(obj, &block)
    res = block.call(obj)
    if res
      res
    elsif obj.supers.empty?
      nil
    else
      obj.supers.find_first do |o|
        lookup(o, &block)
      end
    end
  end

  def self.subclass?(a, b)
    res = subclassb(a, b)
    res
  end
  
  def self.subclassb(a, b)
    an = a.is_a?(String) ? a : a.name
    bn = b.is_a?(String) ? b : b.name
    if a.nil? || b.nil?
      false
    elsif (an == bn)
      true
    else 
      a.supers.any? do |sup|
        subclassb(sup, b)
      end
    end
  end
  
  def self.class_minimum(a, b)
    if b.nil?
      a 
    elsif a.nil?
      b
    elsif subclass?(a, b)
      a
    elsif subclass?(b, a)
      b
    else
      nil
    end
  end
  
  def self.map(obj, &block)
    if obj.nil?
      nil
    else
      res = block.call(obj)
      obj.schema_class.fields.each do |f|
        if f.traversal and !f.type.is_a?("Primitive")
          if !f.many
            map(obj[f.name], &block)
          else
            res[f.name].keys.each do |k|
              map(obj[f.name][k], &block)
            end
          end
        end
      end
      res
    end
  end
end

File: src/js/core/system/library/schema.js
'use strict'

//// Schema ////

var cwd = process.cwd() + '/';
var Enso = require(cwd + "enso.js");

var Schema;

var object_key = function(obj) {
  var self = this;
  return obj.get$(obj.schema_class().key().name());
};

var lookup = function(block, obj) {
  var self = this, res;
  res = block(obj);
  if (res) {
    return res;
  } else if (obj.supers().empty_P()) {
    return null;
  } else {
    return obj.supers().find_first(function(o) {
      return Schema.lookup(block, o);
    });
  }
};

var subclass_P = function(a, b) {
  var self = this, res;
  res = Schema.subclassb(a, b);
  return res;
};

var subclassb = function(a, b) {
  var self = this, an, bn;
  an = Enso.System.test_type(a, String)
    ? a
    : a.name();
  bn = Enso.System.test_type(b, String)
    ? b
    : b.name();
  if (a == null || b == null) {
    return false;
  } else if (an == bn) {
    return true;
  } else {
    return a.supers().any_P(function(sup) {
      return Schema.subclassb(sup, b);
    });
  }
};

var class_minimum = function(a, b) {
  var self = this;
  if (b == null) {
    return a;
  } else if (a == null) {
    return b;
  } else if (Schema.subclass_P(a, b)) {
    return a;
  } else if (Schema.subclass_P(b, a)) {
    return b;
  } else {
    return null;
  }
};

var map = function(block, obj) {
  var self = this, res;
  if (obj == null) {
    return null;
  } else {
    res = block(obj);
    obj.schema_class().fields().each(function(f) {
      if (f.traversal() && ! Enso.System.test_type(f.type(), "Primitive")) {
        if (! f.many()) {
          return Schema.map(block, obj.get$(f.name()));
        } else {
          return res.get$(f.name()).keys().each(function(k) {
            return Schema.map(block, obj.get$(f.name()).get$(k));
          });
        }
      }
    });
    return res;
  }
};

Schema = {
  object_key: object_key,
  lookup: lookup,
  subclass_P: subclass_P,
  subclassb: subclassb,
  class_minimum: class_minimum,
  map: map,
};
module.exports = Schema ;
